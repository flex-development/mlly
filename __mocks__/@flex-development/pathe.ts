/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export default {
  basename: vi.fn(actual.basename),
  changeExt: vi.fn(actual.changeExt),
  dirname: vi.fn(actual.dirname),
  extname: vi.fn(actual.extname),
  format: vi.fn(actual.format),
  isAbsolute: vi.fn(actual.isAbsolute),
  join: vi.fn(actual.join),
  parse: vi.fn(actual.parse),
  relative: vi.fn(actual.relative),
  removeExt: vi.fn(actual.removeExt),
  resolve: vi.fn(actual.resolve)
}
