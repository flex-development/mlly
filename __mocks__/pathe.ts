/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

/**
 * [`pathe`][1] module type.
 *
 * [1]: https://github.com/unjs/pathe
 */
type Actual = typeof import('pathe')

/**
 * `pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const isAbsolute = vi.fn(actual.isAbsolute)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)
