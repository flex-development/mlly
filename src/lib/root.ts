/**
 * @file root
 * @module mlly/lib/root
 */

import { dot, parse, pathToFileURL, resolve } from '@flex-development/pathe'

/**
 * URL of file system root.
 *
 * @const {URL} root
 */
const root: URL = pathToFileURL(parse(resolve(dot)).root)

export default root
