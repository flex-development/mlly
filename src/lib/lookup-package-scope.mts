/**
 * @file lookupPackageScope
 * @module mlly/lib/lookupPackageScope
 */

import isFile from '#lib/is-file'
import root from '#lib/root'
import type { FileSystem, ModuleId } from '@flex-development/mlly'
import pathe from '@flex-development/pathe'

/**
 * Get the package scope URL for the given module `url`.
 *
 * Implements the `LOOKUP_PACKAGE_SCOPE` algorithm.
 *
 * @see {@linkcode FileSystem}
 * @see {@linkcode ModuleId}
 *
 * @async
 *
 * @param {ModuleId} url
 *  URL of module to get package scope for
 * @param {ModuleId | null | undefined} [end]
 *  URL of directory to end search, defaults to {@linkcode root}
 * @param {FileSystem | null | undefined} [fs]
 *  File system API
 * @return {Promise<URL | null>}
 *  URL of nearest directory containing `package.json` file or `null`
 */
async function lookupPackageScope(
  url: ModuleId,
  end?: ModuleId | null | undefined,
  fs?: FileSystem | null | undefined
): Promise<URL | null> {
  /**
   * Scope URL.
   *
   * @var {URL} scopeUrl
   */
  let scopeUrl: URL = new URL(url)

  while (String(scopeUrl) !== String(end ?? root)) {
    // set scopeUrl to URL of parent directory
    scopeUrl = /[/\\]$/.test(scopeUrl.pathname)
      ? new URL(pathe.dot + pathe.dot, scopeUrl)
      : new URL(pathe.dot, scopeUrl)

    // exit early if scopeUrl ends in "node_modules" path segment
    if (/node_modules[/\\]$/.test(scopeUrl.pathname)) break

    // return scopeUrl if `package.json` file exists in directory
    if (await isFile(new URL('package.json', scopeUrl), fs)) return scopeUrl
  }

  return null
}

export default lookupPackageScope
