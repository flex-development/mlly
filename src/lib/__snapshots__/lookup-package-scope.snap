// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:lib/lookupPackageScope > should return URL of package directory (0) 1`] = `file://\${process.cwd()}/`;

exports[`unit:lib/lookupPackageScope > should return URL of package directory (1) 1`] = `file://\${process.cwd()}/`;

exports[`unit:lib/lookupPackageScope > should return URL of package directory (2) 1`] = `file://\${process.cwd()}/`;

exports[`unit:lib/lookupPackageScope > should return URL of package directory (3) 1`] = `file://\${process.cwd()}/node_modules/@commitlint/cli/`;

exports[`unit:lib/lookupPackageScope > should return URL of package directory (4) 1`] = `file://\${process.cwd()}/node_modules/@commitlint/cli/node_modules/@commitlint/config-validator/`;

exports[`unit:lib/lookupPackageScope > should return URL of package directory (5) 1`] = `file://\${process.cwd()}/node_modules/esbuild/`;
